### PR TITLE
feat(genesis bridge): genesis transfers

### PR DESCRIPTION
**Pull request by danwt**
# PR Standards

See

https://github.com/dymensionxyz/research/issues/278 for the master issue.

There are a few outstanding items, that this PR does not try to solve.

This PR

- removes concept of whitelisted genesis triggerer
- adds genesis accounts to the hub genesis rollapp state
- adds an ibc module to send tokens on OnChanOpenConfirm to the hub, containing a special memo which the hub will respect to skip eibc
- adds an ics4 wrapper to stop any user from using the special memo

## Opening a pull request should be able to meet the following requirements

---

Closes https://github.com/dymensionxyz/dymension-rdk/issues/448

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow-up issues.

For Author:

- [ ]  Targeted PR against correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to Github issue with discussion and accepted design
- [ ]  Targets only one github issue
- [ ]  Wrote unit and integration tests
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)

---

For Reviewer:

- [ ]  confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  confirmed all author checklist items have been addressed

---

After reviewer approval:

- [ ]  In case targets main branch, PR should be squashed and merged.
- [ ]  In case PR targets a release branch, PR should be rebased.

## Summary by CodeRabbit

- **New Features**
  - Introduced `IBCModule` for handling channel opening confirmations and genesis account transfers.

- **Improvements**
  - Enhanced the `State` validation logic with specific checks for `Amount` and `Address`.

- **Refactor**
  - Updated module configurations and permissions for better alignment with the latest specifications.
  - Renamed several components for consistency and clarity.
  - Removed redundant functionalities to streamline the codebase.

### CODE REVIEW

File: proto/hub-genesis/params.proto [hunk 1]
```diff
@@ -8,18 +8,6 @@ import "gogoproto/gogo.proto";
 // Params holds parameters for the hub-genesis module.
 message Params {
   option (gogoproto.goproto_stringer) = false;
```
Comment (zale144) [inline proto/hub-genesis/params.proto:10]: could we not just remove `Params` altogether?

File: x/hub-genesis/keeper/genesis_test.go [hunk 2]
```diff
@@ -1,6 +1,7 @@
 package keeper_test
 
 import (
```
Comment (zale144) [inline x/hub-genesis/keeper/genesis_test.go:3]: `goimports`

File: x/hub-genesis/keeper/ibc_module.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
+package keeper
+
+import (
+	"fmt"
```
Comment (zale144) [inline x/hub-genesis/keeper/ibc_module.go:4]: you should install a file save listener that automatically does `goimports`

File: x/hub-genesis/keeper/ibc_module.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+	}
+
+	m := transfertypes.MsgTransfer{
+		SourcePort:       portID,
+		SourceChannel:    channelID,
+		Token:            a.Amount,
+		Sender:           srcAddr,
+		Receiver:         a.GetAddress(),
+		TimeoutHeight:    clienttypes.Height{},
+		TimeoutTimestamp: uint64(ctx.BlockTime().Add(time.Hour * 24).UnixNano()), // TODO: value?
```
Comment (zale144) [inline x/hub-genesis/keeper/ibc_module.go:103]: magic number? pls const it

File: x/hub-genesis/keeper/memo.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+	errors2 "github.com/cosmos/cosmos-sdk/types/errors"
+	types2 "github.com/cosmos/cosmos-sdk/x/bank/types"
+)
+
+func memoHasKey(memo string) bool {
+	m := make(map[string]any)
+	if err := json.Unmarshal([]byte(memo), &m); err != nil {
+		return false
+	}
+	_, ok := m["genesis_transfer"]
```
Comment (zale144) [inline x/hub-genesis/keeper/memo.go:18]: const the key

File: x/hub-genesis/keeper/memo.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+	return ok
+}
+
+// createMemo creates a memo to go with the transfer. It's used by the hub to confirm
+// that the transfer originated from the chain itself, rather than a user of the chain.
+// It may also contain token metadata.
+func (w IBCModule) createMemo(ctx types.Context, denom string, i, n int) (string, error) {
+	d, ok := w.getDenom(ctx, denom)
+	if !ok {
+		return "", errors.Wrap(errors2.ErrNotFound, "get denom metadata")
```
Comment (zale144) [inline x/hub-genesis/keeper/memo.go:28]:
```suggestion
		return "", errors.Wrap(sdkerrors.ErrNotFound, "get denom metadata")
```

File: x/hub-genesis/keeper/memo.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+	if err != nil {
+		return "", errors2.ErrJSONMarshal
+	}
+
+	return string(bz), nil
+}
+
+type memo struct {
+	Data struct {
+		Denom types2.Metadata `json:"denom"`
```
Comment (zale144) [inline x/hub-genesis/keeper/memo.go:46]:
```suggestion
		Denom banktypes.Metadata `json:"denom"`
```

File: x/hub-genesis/keeper/params.go [hunk 8]
```diff
@@ -15,19 +15,3 @@ func (k Keeper) GetParams(ctx sdk.Context) (params types.Params) {
 func (k Keeper) SetParams(ctx sdk.Context, params types.Params) {
```
Comment (zale144) [inline x/hub-genesis/keeper/params.go:15]: again, Params is now empty, can we not completely get rid of it for this module? Will a migration be needed?

File: x/hub-genesis/types/codec.go [hunk 9]
```diff
@@ -4,25 +4,17 @@ import (
...
 )
 
 func RegisterCodec(cdc *codec.LegacyAmino) {
-	cdc.RegisterConcrete(&MsgHubGenesisEvent{}, "hub-genesis/HubGenesisEvent", nil)
 }
 
 func RegisterInterfaces(registry cdctypes.InterfaceRegistry) {
-	registry.RegisterImplementations((*sdk.Msg)(nil), &MsgHubGenesisEvent{})
-	msgservice.RegisterMsgServiceDesc(registry, &_Msg_serviceDesc)
 }
```
Comment (zale144) [inline x/hub-genesis/types/codec.go:13]: remove?

File: x/hub-genesis/types/keys.go [hunk 10]
```diff
@@ -13,7 +12,7 @@ const (
 	// QuerierRoute is the querier route for the minting store.
 	QuerierRoute = StoreKey
 
-	// Query endpoints supported by the minting querier.
+	// QueryParameters endpoints supported by the minting querier.
```
Comment (zale144) [inline x/hub-genesis/types/keys.go:15]: `QueryParameters` is not used

Comment (zale144) [reply x/hub-genesis/types/keys.go:15]: also `RouterKey` is not used

File: x/hub-genesis/types/params.go [hunk 11]
```diff
@@ -1,31 +1,23 @@
 package types
```
Comment (zale144) [inline x/hub-genesis/types/params.go:1]: as above, we should be able to remove this file and all `Params` related stuff

**Review by zale144 — COMMENTED** [review]
nice work! left some comments

File: x/hub-genesis/keeper/memo.go [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+func (w IBCModule) createMemo(ctx types.Context, denom string, i, n int) (string, error) {
+	d, ok := w.getDenom(ctx, denom)
+	if !ok {
+		return "", errors.Wrap(errors2.ErrNotFound, "get denom metadata")
+	}
+
+	m := memo{}
+	m.Data.Denom = d
+	m.Data.TotalNumTransfers = uint64(n)
+	m.Data.ThisTransferIx = uint64(i)
```
Comment (zale144) [inline x/hub-genesis/keeper/memo.go:34]: "Ix" is like "Idx"?

File: x/hub-genesis/keeper/memo.go (hunk 5 above)
Comment (danwt) [reply x/hub-genesis/keeper/memo.go:18]: done

File: x/hub-genesis/keeper/ibc_module.go (hunk 4 above)
Comment (danwt) [reply x/hub-genesis/keeper/ibc_module.go:103]: ahh good spot

File: x/hub-genesis/keeper/memo.go (hunk 6 above)
Comment (danwt) [reply x/hub-genesis/keeper/memo.go:28]: sorry good spot, it was from auto refac

File: x/hub-genesis/keeper/memo.go (hunk 7 above)
Comment (danwt) [reply x/hub-genesis/keeper/memo.go:46]: sorry good spot, it was from auto refac

File: x/hub-genesis/keeper/memo.go (hunk 12 above)
Comment (danwt) [reply x/hub-genesis/keeper/memo.go:34]: yes, pretty standard

File: x/hub-genesis/types/codec.go (hunk 9 above)
Comment (danwt) [reply x/hub-genesis/types/codec.go:13]: Yes you're right it should be removed
but is it ok if I do it in a followup PR? This PR is currently all tested etc
https://github.com/dymensionxyz/dymension-rdk/issues/452
I have made an issue for myself
Thanks

File: proto/hub-genesis/params.proto (hunk 1 above)
Comment (danwt) [reply proto/hub-genesis/params.proto:10]: Yes you're right it should be removed
but is it ok if I do it in a followup PR? This PR is currently all tested etc
https://github.com/dymensionxyz/dymension-rdk/issues/452
I have made an issue for myself
Thanks

File: x/hub-genesis/keeper/params.go (hunk 8 above)
Comment (danwt) [reply x/hub-genesis/keeper/params.go:15]: Yes you're right it should be removed
but is it ok if I do it in a followup PR? This PR is currently all tested etc
https://github.com/dymensionxyz/dymension-rdk/issues/452
I have made an issue for myself
Thanks

File: x/hub-genesis/types/keys.go (hunk 10 above)
Comment (danwt) [reply x/hub-genesis/types/keys.go:15]: ditto

Yes you're right it should be removed
but is it ok if I do it in a followup PR? This PR is currently all tested etc
https://github.com/dymensionxyz/dymension-rdk/issues/452
I have made an issue for myself
Thanks

File: x/hub-genesis/types/params.go (hunk 11 above)
Comment (danwt) [reply x/hub-genesis/types/params.go:1]: Yes you're right it should be removed
but is it ok if I do it in a followup PR? This PR is currently all tested etc
https://github.com/dymensionxyz/dymension-rdk/issues/452
I have made an issue for myself
Thanks

File: x/hub-genesis/keeper/ibc_module.go (hunk 4 above)
Comment (danwt) [reply x/hub-genesis/keeper/ibc_module.go:103]: I am working on a fix now

Comment (danwt) [reply x/hub-genesis/keeper/ibc_module.go]: fixed

**Review by srene — APPROVED** [review]

**Comment by danwt** [comment]
Thanks @srene 
I just fixed the validate func
Please could you green again?

**Review by srene — APPROVED** [review]

**Review by srene — APPROVED** [review]

**Review by zale144 — APPROVED** [review]
lgtm

**Comment by mtsitrin** [comment]
merged in the middle of review

File: x/hub-genesis/types/state.go [hunk 13]
```diff
@@ -1,9 +1,19 @@
...
+import (
+	errorsmod "cosmossdk.io/errors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
 func (s State) Validate() error {
-	if !s.GenesisTokens.IsValid() {
-		return ErrInvalidGenesisTokens
+	for _, a := range s.GetGenesisAccounts() {
+		if err := a.GetAmount().Validate(); err != nil {
```
Comment (mtsitrin) [inline x/hub-genesis/types/state.go:10]: validate denom is the native denom?

File: x/hub-genesis/keeper/ibc_module.go [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+func (w IBCModule) mintAndTransfer(
+	ctx sdk.Context,
+	n int,
+	a types.GenesisAccount,
+	srcAddr string,
+	portID string,
+	channelID string,
+) error {
+	coin := a.GetAmount()
+	err := w.mintCoins(ctx, types.ModuleName, sdk.Coins{coin})
```
[the remark this answers is not in the document]
Comment (zale144) [reply x/hub-genesis/keeper/ibc_module.go:88]: shouldn't we also support rollapps operating with exclusively 3rd party tokens?

File: x/hub-genesis/keeper/ibc_module.go [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+	k         Keeper
+	transfer  Transfer
+	getDenom  GetDenomMetaData
+	mintCoins MintCoins
+}
+
+type (
+	Transfer         func(ctx sdk.Context, transfer *transfertypes.MsgTransfer) error
+	GetDenomMetaData func(ctx sdk.Context, denom string) (banktypes.Metadata, bool)
+	MintCoins        func(ctx sdk.Context, moduleName string, amt sdk.Coins) error
```
[the remark this answers is not in the document]
Comment (zale144) [reply x/hub-genesis/keeper/ibc_module.go:32]: I'm not saying we should strive towards SDK's shitty standards. Just that our codebase should have consistent coding style in every module. As for great innovation, keeping it simple and easy to understand is enough of a challenge on its own.

File: x/hub-genesis/types/state.go (hunk 13 above)
Comment (danwt) [reply x/hub-genesis/types/state.go:10]: discuss here https://github.com/dymensionxyz/dymension-rdk/pull/449#discussion_r1647542987

File: x/hub-genesis/keeper/genesis_test.go [hunk 16]
```diff
@@ -1,6 +1,7 @@
...
+	_ "embed"
```
[the remark this answers is not in the document]
Comment (danwt) [reply x/hub-genesis/keeper/genesis_test.go:4]: see https://github.com/dymensionxyz/dymension-rdk/pull/458

File: x/hub-genesis/keeper/ibc_module.go [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+
+	for i, a := range state.GetGenesisAccounts() {
+		if err := w.mintAndTransfer(ctx, len(state.GetGenesisAccounts()), a, srcAddr, portID, channelID); err != nil {
+			// there is no feasible way to recover
+			panic(fmt.Errorf("mint and transfer: %w", err))
+		}
+		l.Info("Sent genesis transfer.", "index", i, "receiver", a.GetAddress(), "coin", a)
+	}
+
+	state.GenesisAccounts = nil
```
[the remark this answers is not in the document]
Comment (danwt) [reply x/hub-genesis/keeper/ibc_module.go:70]: see https://github.com/dymensionxyz/dymension-rdk/pull/458

File: x/hub-genesis/keeper/ibc_module.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (danwt) [reply x/hub-genesis/keeper/ibc_module.go:88]: see https://github.com/dymensionxyz/dymension-rdk/pull/458
https://github.com/dymensionxyz/dymension-rdk/issues/462

**Pull request merged**